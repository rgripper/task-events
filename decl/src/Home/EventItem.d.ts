import * as React from "react";
import * as TMEvents from "../api/TMEvents";
interface Props {
    event: TMEvents.Event;
}
export declare const EventItem: React.SFC<Props>;
export {};
