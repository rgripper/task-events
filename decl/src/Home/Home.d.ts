import * as React from "react";
import { UserSettingsState } from "../store/models";
import * as TMEvents from "../api/TMEvents";
interface MappedStateProps {
    events: TMEvents.Event[];
    isSearching: boolean;
    userSettings: UserSettingsState;
}
declare type Props = MappedStateProps;
export declare class _Home extends React.Component<Props> {
    render(): JSX.Element;
}
export declare const Home: import("react-redux").ConnectedComponentClass<typeof _Home, Pick<MappedStateProps, never>>;
export {};
