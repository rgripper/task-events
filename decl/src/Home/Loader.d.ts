import * as React from "react";
export declare const Container: import("create-emotion-styled").StyledOtherComponent<{}, React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>, any>;
export declare const Loader: () => JSX.Element;
