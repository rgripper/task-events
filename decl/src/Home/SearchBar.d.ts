import * as React from "react";
interface MappedStateProps {
    value: string;
}
interface MappedDispatchProps {
    onChange: (value: string) => void;
}
declare type Props = MappedStateProps & MappedDispatchProps;
declare class _SearchBar extends React.Component<Props> {
    memoizeDebounce: (fun: any) => any;
    render(): JSX.Element;
}
export declare const SearchBar: import("react-redux").ConnectedComponentClass<typeof _SearchBar, Pick<Props, never>>;
export {};
