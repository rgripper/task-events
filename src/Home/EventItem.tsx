import * as React from "react";
import * as TMEvents from "../api/TMEvents";
import styled from "react-emotion";

interface Props {
  event: TMEvents.Event;
}

const Container = styled("a")`
  color: #333;
  text-decoration: none;
  position: relative;
  height: 100%;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  display: flex;
  flex-direction: column;
  cursor: pointer;

  &:hover {
    box-shadow: 0px 4px 4px rgba(124, 124, 0, 0.25);
  }

  > img {
    flex: 1;
  }
`;

const ThumbImage = styled("img")`
  height: 14rem;
  object-fit: cover;

  &[src=""] {
    background: #eee;
  }
`;

const Price = styled("span")`
  position: absolute;
  top: 0.5rem;
  left: 0.5rem;
  padding: 0.25rem;
  border-radius: 0.5rem;
  background-color: #ccc;
  opacity: 0.75;
`;

const Description = styled("div")`
  display: flex;
  align-items: center;
  font-size: 1.125rem;
  padding: 0.5rem;
  height: 3rem;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const getThumbImage = (images: TMEvents.Image[]): TMEvents.Image | undefined =>
  images.filter(x => x.width > 300).sort((a, b) => a.width - b.width)[0];

export const EventItem: React.SFC<Props> = props => {
  const image = getThumbImage(props.event.images);
  const priceRange = props.event.priceRanges && props.event.priceRanges[0];
  return (
    <Container target="_blank" href={props.event.url}>
      {priceRange && (
        <Price>
          {priceRange.min} - {priceRange.max} AUD
        </Price>
      )}
      <ThumbImage src={image ? image.url : ""} />
      <Description>{props.event.name}</Description>
    </Container>
  );
};
