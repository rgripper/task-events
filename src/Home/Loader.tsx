import styled from "react-emotion";
import * as React from "react";

export const Container = styled("div")`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 0.125rem;
  width: 8rem;
  height: 8rem;

  > div {
    position: relative;
    display: inline-block;
    width: 100%;
    height: 100%;
    background: #ccc;
    transform: scale(0);
    transform-origin: center center;
    animation: loader 2s infinite linear;

    &:nth-of-type(7) {
    }

    &:nth-of-type(1),
    &:nth-of-type(5),
    &:nth-of-type(9) {
      animation-delay: 0.4s;
    }

    &:nth-of-type(4),
    &:nth-of-type(8) {
      animation-delay: 0.2s;
    }

    &:nth-of-type(2),
    &:nth-of-type(6) {
      animation-delay: 0.6s;
    }

    &:nth-of-type(3) {
      animation-delay: 0.8s;
    }
  }

  @keyframes loader {
    0% {
      transform: scale(0);
    }
    40% {
      transform: scale(1);
    }
    80% {
      transform: scale(1);
    }
    100% {
      transform: scale(0);
    }
  }
`;

export const Loader = () => (
  <Container>
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </Container>
);
