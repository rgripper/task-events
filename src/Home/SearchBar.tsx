import * as React from "react";
import { RootState } from "../store/models";
import { connect } from "react-redux";
import memoizeOne from "memoize-one";
import { debounce } from "throttle-debounce";
import styled from 'react-emotion';

interface MappedStateProps {
  value: string;
}

interface MappedDispatchProps {
  onChange: (value: string) => void;
}

type Props = MappedStateProps & MappedDispatchProps;

const Container = styled('div')`
  display: flex;
  border: 0.075rem solid #ccc;
  padding: 0.5rem;
  border-radius: 0.5rem;
  align-items: center;
  color: #ccc;

  > i {
    margin-right: 0.5rem;
  }

  > input {
    flex: 1;
    border: 0;
    outline: 0;
    font-size: 1.125rem;
  }
`;

class _SearchBar extends React.Component<Props> {

  // make sure we cache and use the same debouncer
  memoizeDebounce = memoizeOne(fun => debounce(300, fun));

  render() {
    return (
      <Container>
        <i className="fa fa-search" />
        <input
          type="text"
          placeholder="Search keyword, e.g. music..."
          onChange={event => this.memoizeDebounce(this.props.onChange)(event.currentTarget.value)}
        />
      </Container>
    );
  }
}

export const SearchBar = connect(
  (state: RootState): MappedStateProps => ({
    value: state.events.keyword
  }),
  ({ events }: any): MappedDispatchProps => ({
    onChange: (keyword: string) => events.load(keyword)
  })
)(_SearchBar);