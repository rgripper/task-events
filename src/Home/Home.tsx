import * as React from "react";
import { connect } from "react-redux";
import { RootState, UserSettingsState } from "../store/models";
import * as TMEvents from "../api/TMEvents";
import { EventItem } from "./EventItem";
import { SearchBar } from "./SearchBar";
import styled from "react-emotion";
import { Loader } from "./Loader";

interface MappedStateProps {
  events: TMEvents.Event[];
  isSearching: boolean;
  userSettings: UserSettingsState;
}

type Props = MappedStateProps;

const EventListContainer = styled("div")`
  padding: 1rem;
  display: grid;
  grid-template-columns: repeat(auto-fill, 19rem);
  grid-gap: 1rem;
  align-items: center;
  justify-content: center;

  > {
    margin: 0 auto;
    max-width: 60rem;
  }
`;

const Center = styled("div")`
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const SearchBarContainer = styled("div")`
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  padding: 2rem;

  > {
    margin: 0 auto;
    max-width: 60rem;
  }
`;

const Container = styled("div")`
  height: 100vh;
  display: flex;
  flex-direction: column;
`;

const NoResults = styled("div")`
  margin: 2rem;
  text-align: center;
  color: #ccc;
  font-size: 1.125rem;
`;

const SearchHint = styled("small")`
  color: #ccc;
  font-size: 0.75rem;
  padding: 0.5rem 0;
`;

export class _Home extends React.Component<Props> {
  public render() {
    return (
      <Container>
        <SearchBarContainer>
          <SearchBar />
          <SearchHint>
            {this.props.userSettings.coordinates
              ? `Search around coordinates: ${
                  this.props.userSettings.coordinates.latitude
                }, ${this.props.userSettings.coordinates.longitude}`
              : `Search in country: ${this.props.userSettings.countryCode}`}
          </SearchHint>
        </SearchBarContainer>

        {this.props.isSearching ? (
          <Center>
            <Loader />
          </Center>
        ) : this.props.events.length > 0 ? (
          <div>
            <EventListContainer>
              {this.props.events.map(event => (
                <EventItem key={event.id} event={event} />
              ))}
            </EventListContainer>
          </div>
        ) : (
          <Center>
            <NoResults>No events found so far.</NoResults>
          </Center>
        )}
      </Container>
    );
  }
}

export const Home = connect(
  (state: RootState): MappedStateProps => ({
    events: state.events.items,
    isSearching: state.events.isSearching,
    userSettings: state.userSettings
  })
)(_Home);
