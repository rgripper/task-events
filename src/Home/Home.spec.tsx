import * as React from "react";
import { shallow } from "enzyme";
import { _Home } from "./Home";

describe("<Home />", () => {
  const setup = () => {
    return shallow(<_Home isSearching={false} events={[]} />);
  };

  it("should render successfully", () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
