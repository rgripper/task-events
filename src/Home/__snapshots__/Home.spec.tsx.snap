// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render successfully 1`] = `
<Styled(div)>
  <Styled(div)>
    <Connect(_SearchBar) />
    <Styled(small)>
      Search in country: AU
    </Styled(small)>
  </Styled(div)>
  <Styled(div)>
    <Styled(div)>
      No events found so far.
    </Styled(div)>
  </Styled(div)>
</Styled(div)>
`;
